### PR TITLE
Feature/update cloud example

### LINKED ISSUE #342: Update cloud example with re-imaging
The cloud with re-imaging example needs to be updated to incorporate new network architecture and observation space.

### PR DESCRIPTION
**Pull request by LorenzzoQM**
## Description
Closes #342 

Updates the cloud with reimaging example script to incorporate recent changes and new actor-critic network architecture.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### How should this pull request be reviewed?
- [ ] By commit
- [X] All changes at once

## Future Work

Update with the related publication information once available.

## Checklist

- [X] I have performed a self-review of my code
- [X] I have commented my code in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and release notes
- [X] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: examples/cloud_environment_with_reimaging.ipynb [hunk 1]
```diff
@@ -685,68 +684,154 @@
...
+    "class CS_F(sats.ImagingSatellite):\n",
     "    observation_spec = [\n",
     "        obs.SatProperties(\n",
-    "            dict(prop=\"omega_BP_P\", norm=0.03),\n",
-    "            dict(prop=\"c_hat_P\"),\n",
-    "            dict(prop=\"r_BN_P\", norm=orbitalMotion.REQ_EARTH * 1e3),\n",
-    "            dict(prop=\"v_BN_P\", norm=7616.5),\n",
-    "            dict(prop=\"battery_charge_fraction\"),\n",
-    "            dict(prop=\"solar_angle_norm\"),\n",
+    "            dict(prop=\"omega_BN_B\", norm=0.03),  # 0,1,2 - CHECKED\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Those "CHECKED" comments should be removed.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 2]
```diff
@@ -850,7 +942,7 @@
     "    sim_rate=0.5,\n",
     "    max_step_duration=300.0,\n",
     "    time_limit=95 * 60 / 2,  # half orbit\n",
-    "    log_level=\"INFO\",\n",
+    "    log_level=\"WARNING\",\n",
     "    failure_penalty=0.0,\n",
     "    # disable_env_checker=True,  # For debugging\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: I think the "disable_env_checker" doesn't exist anymore and we need to remove this.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 3]
```diff
@@ -959,13 +1261,732 @@
...
+    "    target_pos = np.argmax(list_scores)\n",
+    "\n",
+    "    return target_pos + 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training and custom actor-critic modules\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:1344]: I feel like this code is too much for one example script. Have you thought about splitting the code into two? For cloud with re-imaging example. And training with new architecture script?

File: examples/cloud_environment_with_reimaging.ipynb
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:57]: Add explanation for "belief_update_var"

File: examples/cloud_environment_with_reimaging.ipynb [hunk 4]
```diff
@@ -312,8 +312,8 @@
     "\n",
     "$$\n",
     "R = \\begin{cases}\n",
-    "\\rho_i\\alpha_i\\Delta \\text{P}(S=1) + \\rho_i(1 - \\alpha) & \\text{ if }  \\text{P}_i(S=1) \\geq \\theta_{\\text{thr}_i} \\\\\n",
-    "\\rho_i\\alpha_i\\Delta \\text{P}(S=1) & \\text{ otherwise.}\n",
+    "r_i\\alpha_i\\Delta \\text{P}(S=1) + r_i(1 - \\alpha) & \\text{ if }  \\text{P}_i(S=1) \\geq \\theta_{\\text{thr}_i} \\\\\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: The alpha and alpha_i the same, right? If so, make it consistent. If not, explain the difference.

File: examples/cloud_environment_with_reimaging.ipynb
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Doesn't this have to be "CloudImageProbabilityData" and not "UniqueImageData"?

Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Isn't this "+ priority * belief_variation * alpha"? I thought the variation is also applied even when it reaches the threshold.

Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Use consistent parameters as in the equation (e or eta instead of k).

Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Is it 30 mins to get the sigmoid function to be 1? It's not 30 mins just for the equation.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 5]
```diff
@@ -685,68 +684,154 @@
...
+    "        .read()\n",
+    "        .PositionVector\n",
+    "    )\n",
+    "    r_BN_N = sat.dynamics.r_BN_N\n",
+    "    r_SB_N = np.array(r_SN_N) - np.array(r_BN_N)\n",
+    "    r_SB_H = sat.dynamics.HN @ r_SB_N\n",
+    "    return r_SB_H / np.linalg.norm(r_SB_H)\n",
+    "\n",
+    "\n",
+    "class CloudCoverDensity(obs.Observation):\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:711]: Add this class explanation somewhere.

File: examples/cloud_environment_with_reimaging.ipynb
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:674]: It would be great if you could briefly explain what each satellite is (difference between CS_F and CS_F_Bayesian).

File: examples/cloud_environment_with_reimaging.ipynb [hunk 6]
```diff
@@ -685,68 +684,154 @@
...
-    "            solar_panel_vec_B = np.array([0, 0, -1])  # Not default configuration\n",
-    "            mat = np.transpose(self.BN)\n",
-    "            solar_panel_vec_N = np.matmul(mat, solar_panel_vec_B)\n",
-    "            error_angle = np.arccos(np.dot(solar_panel_vec_N, sun_vec_N_hat))\n",
-    "\n",
-    "            return error_angle / np.pi\n",
-    "\n",
-    "    dyn_type = CustomDynModel\n",
-    "    fsw_type = fsw.SteeringImagerFSWModel"
+    "        self.also_image_filter = []"
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: I don't see this "also_image_filter" is used anywhere.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 7]
```diff
@@ -888,10 +980,217 @@
     "print(\"Actions:\", env.satellites[0].action_description, \"\\n\")\n",
     "print(\"States:\", env.unwrapped.satellites[0].observation_description, \"\\n\")\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: I think this should be "Observations" instead of "States"

File: examples/cloud_environment_with_reimaging.ipynb [hunk 8]
```diff
@@ -888,10 +980,217 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Greedy Heuristic\n",
+    "\n",
+    "A greedy heuristic can be created to task the satellite. For the heuristic with $\\alpha=1.0$, a target is selected from the $N=35$ upcoming targets with\n",
+    "$$\n",
+    "\\rho^\\star=\\arg\\max_{\\rho_i \\in \\mathcal{U}_{35}} \\left\\{\\frac{r_i \\Delta\\text{P}_i(S=1)}{t_{i,{\\text{max}}}}  \\; : \\; \\text{P}_i(S=1) < \\theta_{\\text{thr}_i} \\; t_{i,\\text{max}} < t_{i,\\text{close}}\\right\\}\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Explain what t_i,max and t_i,close are

File: examples/cloud_environment_with_reimaging.ipynb [hunk 9]
```diff
@@ -907,20 +1206,23 @@
...
+    "        action_dict = {\n",
+    "            sat_i.name: greedy_heuristic_reimaging(sat_i, 0.0, 35)\n",
+    "            for sat_i in env.satellites\n",
+    "        }\n",
+    "    steps += 1\n",
     "\n",
     "    observation, reward, terminated, truncated, info = env.step(action_dict)\n",
     "\n",
-    "    if all(terminated.values()) or all(truncated.values()):\n",
+    "    if all(terminated.values()) or all(truncated.values()) or steps >= 5:\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:1305]: I think it would be better to see the info message instead of just episode complete output. Is there any reason why you avoid it and made it to "warning" especially since you didn't actually run training in the script?

File: examples/cloud_environment_with_reimaging.ipynb [hunk 10]
```diff
@@ -959,13 +1261,732 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Check [Training with RLlib PPO](../examples/rllib_training.ipynb) for an example on how to train the agent in this environment."
+    "A similar heuristic can be used for the single-picture case, where the expected reward is used instead of the rewards to compute the reward density per target with\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Just the order of the heuristic policy. I think the single image one should be first to be consistent.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 11]
```diff
@@ -959,13 +1261,732 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training and custom actor-critic modules\n",
+    "\n",
+    "The observation space contains concatenated information regarding the spacecraft and individual targets, which naturally decomposes into two components: a spacecraft state vector and a set of target-specific state vectors. This representation motivates the use of shared target encoders, attention mechanisms, and permutation-invariant pooling operations instead of applying an MLP directly to the flattened observation. \n",
+    "\n",
+    "Then, custom actor and critic modules are defined for training with PPO."
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:1350]: Add recommended (or used) ray and torch version for this training. 

Also, did you rewrite all of the code here from the Ray codebase? I assume some parts were copied from other sources and then combined or modified. It would be helpful to clearly show which parts came from the original source and which parts you changed.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 12]
```diff
@@ -959,13 +1261,732 @@
...
+    "        vf_out = self.vf(batch)\n",
+    "        outs[Columns.VF_PREDS] = vf_out.squeeze(-1)\n",
+    "        return outs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Then, the training hyperparameters and modules parameters are defined, as well as the custom RLModule."
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb]: Add short comment that those parameters are chosen by hyperparameter search.

File: examples/cloud_environment_with_reimaging.ipynb [hunk 13]
```diff
@@ -959,13 +1261,732 @@
...
+    "    .framework(framework=\"torch\")\n",
+    "    .api_stack(\n",
+    "        enable_rl_module_and_learner=True,\n",
+    "        enable_env_runner_and_connector_v2=True,\n",
+    "    )\n",
+    "    .callbacks(WrappedEpisodeDataCallbacks)\n",
+    ")\n",
+    "ppo_config.rl_module(**rl_module_args)\n",
+    "\n",
+    "# Uncomment to run training\n",
```
Comment (Yume27) [inline examples/cloud_environment_with_reimaging.ipynb:1976]: ray and tune are not imported which are used in the commented code.

File: examples/cloud_environment_with_reimaging.ipynb (hunk 10 above)
Comment (LorenzzoQM) [reply examples/cloud_environment_with_reimaging.ipynb]: I agree that changing the order would make it more consistent. I also changed the sections to present the greedy heuristic first and then the env initialization and interaction. Let me know if this improved the flow.

File: examples/cloud_environment_with_reimaging.ipynb (hunk 11 above)
Comment (LorenzzoQM) [reply examples/cloud_environment_with_reimaging.ipynb:1350]: The ray and torch versions are the ones specified in the BSK-RL requirements, so it should work by default.

File: examples/cloud_environment_with_reimaging.ipynb (hunk 10 above)
Comment (Yume27) [reply examples/cloud_environment_with_reimaging.ipynb]: Like this new order!

**Review by Yume27 — APPROVED** [review]
Great example and show the new structure! Maybe good to think about splitting the code into two later, but good for now. And passed the tests and ran correctly.
